jsonargparse/_typehints.py

```diff
@@ -904,6 +939,16 @@ def adapt_typehints(
         elif typehint_origin is OrderedDict:
             val = dict(val) if serialize else OrderedDict(val)
 
+    # NotRequired
+    elif typehint_origin in not_required_types:
+        assert len(subtypehints) == 1, "NotRequired requires a single type argument"
+        val = adapt_typehints(val, subtypehints[0], **adapt_kwargs)
```

Comment:
This line is identical for both `NotRequired` and `Required`. Also, this should only be encountered for `TypedDict` items. Maybe better to do this right after `for k, v in val.items():` in line 935 above? Like for the case of TypedDict check if origin is either `NotRequired` or `Required`, then call adapt_typehints with the first argument of the type.

Like this maybe the code looks cleaner, so not sure. But if it is like this, then merge both and use as comment `# TypedDict NotRequired and Required`.

Reply:
Right, I thought it was a bit cleaner to put it here, and I believed it was more future-proof if the semantics of `NotRequired` and `Required` ever expand to contexts beyond `TypedDict`. I'll merge the two branches.